### PR TITLE
Improve ec2_vpc_subnet check mode

### PR DESCRIPTION
**Pull request by willthames**
##### SUMMARY

check_mode should behave pretty similarly to non-check mode -
just don't actually create or delete subnets or change tags.

Using DryRun for check_mode behaves very differently and results
in the following module failure:

```
"msg": "Unable to update tags for subnet-abcd1234,
error: EC2ResponseError: 412 Precondition Failed
<?xml version=\"1.0\" encoding=\"UTF-8\"?>
<Response><Errors><Error><Code>DryRunOperation</Code>
<Message>Request would have succeeded, but DryRun flag
is set.</Message></Error></Errors>
<RequestID>12345678-abcd-1234-abcd-abcd1234abcd</RequestID></Response>"
```

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
ec2_vpc_subnet

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel c1b3d6a51f) last updated 2017/03/30 09:17:50 (GMT +1000)
  config file = ./ansible.cfg
  configured module search path = [u'./library']
  python version = 2.7.13 (default, Jan 12 2017, 17:59:37) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py [hunk 1]
```diff
@@ -168,14 +168,17 @@ def ensure_tags(vpc_conn, resource_id, tags, add_only, check_mode):
...
+            vpc_conn.delete_tags(resource_id, to_delete)
 
         to_add = dict((k, tags[k]) for k in tags if k not in cur_tags or cur_tags[k] != tags[k])
-        if to_add:
-            vpc_conn.create_tags(resource_id, to_add, dry_run=check_mode)
+        if to_add and not check_mode:
+            vpc_conn.create_tags(resource_id, to_add)
 
-        latest_tags = get_resource_tags(vpc_conn, resource_id)
+        if check_mode:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py]: You don't need the if/else block here.

**Review by s-hertel — APPROVED** [review]

File: lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py (hunk 1 above)
Comment (willthames) [reply lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py]: Why's that? My reasoning for the if/else block is that if in check_mode, the resource won't have had its tags updated, so querying the resource's tags will provide incorrect info in check mode. However, out of check mode, it's worth validating what the resource actually thinks its tags are (we could just say that `latest_tags = tags` and not make the API call, but I'd rather be accurate, as that'll expose issues sooner)

Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py]: Oh, you're right.

**Comment by s-hertel** [comment]
shipit

**Comment by willthames** [comment]
Now that #23783 is merged, my tests on this are out of date. If anyone can test I've not introduced any further errors, that would be good, otherwise I'll update this once I've retested.

**Comment by willthames** [comment]
cc @brandond

**Comment by willthames** [comment]
Retested, all looks good to merge to me.

**Review by s-hertel — APPROVED** [review]

**Comment by s-hertel** [comment]
shipit

**Pull request merged**